Extend pipes to support use from ISR

peter-mitsis opened the pull request:
This set of commits updates the pipes implementation and its documentation.

It has a few notable differences as compared to the previous implementation.

1. The use of schedule locking has been eliminated in favour of spinlocks. This both makes it safer for SMP and allows pipes to be used from the context of an ISR. This comes at the price of increasing interrupt latency while copying large amounts of data to/from the pipe.
2. Instead of using custom copy loops depending upon the location of source and destination buffers, I've tried to unify things more so that copies are done using a sort of scatter-gather. When both writing to and reading from the pipe, we set up descriptors that track where the data is going/coming from as well as how much is list. The copying then just loops over these descriptors processing them one by one.
3. The manner in which the "working list" is generated and linked has also been simplified. Instead of re-purposing the thread's queuing node to link the list together, the descriptors are linked together using k_pipe_desc instances on the stack.
    
Fixes #47061

peter-mitsis committed:
kernel: pipes: rewrite pipes implementation

This new implementation of pipes has a number of advantages over the
previous.
  1. The schedule locking is eliminated both making it safer for SMP
     and allowing for pipes to be used from ISR context.
  2. The code used to be structured to have separate code for copying
     to/from a wating thread's buffer and the pipe buffer. This had
     unnecessary duplication that has been replaced with a simpler
     scatter-gather copy model.
  3. The manner in which the "working list" is generated has also been
     simplified. It no longer tries to use the thread's queuing node.
     Instead, the k_pipe_desc structure (whose instances are on the
     part of the k_thread structure) has been extended to contain
     additional fields including a node for use with a linked list. As
     this impacts the k_thread structure, pipes are now configurable
     in the kernel via CONFIG_PIPES.

Fixes #47061

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

peter-mitsis committed:
doc: Update documentation related to pipes

The pipes implementation has been updated to allow ISRs to both
send and receive data to/from pipes provided the K_NO_WAIT timeout
option is used.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

mbolivar-nordic commented on doc/kernel/services/data_passing/pipes.rst:
```diff
@@ -54,8 +54,9 @@ pended writer to fill the ring buffer.
 .. note::
-    The kernel does NOT allow for an ISR to send or receive data to/from a
-    pipe or flush even if it does not attempt to wait for space/data.
+    The kernel does allow for an ISR to flush a pipe from an ISR. It also
```
It would be nice to preemptively add this to doc/releases/release-notes-3.2.rst to make sure we don't forget to note the change in the release crush. Not a blocking comment.

peter-mitsis replied:
I'll make a note of that locally. Updating the 3.2 release notes is something that I would like to do after this gets merged (if I may be so presumptuous) to reduce the likelihood of merge conflicts.

andyross reviewed (commented):
Did a quick first-pass review and can't find anything wrong (or really any notable nitpicks, even).

One thing that would be nice, given that our existing test coverage (and, to be fair, app uptake) on k_pipe is a little thin, would be a multithreaded "stress" case that hammers the pipe form ISR and SMP contexts looking for edges and glitches.

andyross commented on kernel/pipes.c:
```diff
@@ -557,173 +476,156 @@ static int pipe_get_internal(k_spinlock_key_t key, struct k_pipe *pipe,
-	 *    call is complete so it is still safe to copy data from it.
-	 */
+	src_desc = (struct k_pipe_desc *)sys_dlist_get(&src_list);
+	while (src_desc != NULL) {
```
The logic in this loop is almost but not quite identical to the same copy loop in k_pipe_put(), and I'm wondering if it's worth further unifying the treatment.  Both schemes are just iterating over two lists of buffers to copy N bytes between them, it's just that one side of each version is guaranteed to have only one element (the buffer provided to the API by the caller).

But beyond that?  They're the same code.  And the "only one element" distinction is something that can be handled with just one extra test at the top of the loop (and an extra dummy list head somewhere).

I love the semantic unification of this PR, but it does seem like there's more fruit on the tree.

peter-mitsis replied:
Let's see what I can do. :)

peter-mitsis replied:
On the surface, there are a lot of similarities in the two loops. So much so that it does seem natural enough to want to unify them. However, there are a substantial number of minor differences that stem from the direction that data flows (adding vs subtraction, whether its a source or destination descriptor's thread that must be NULL, ...) . With the current proposal, my initial attempts just generate a hacky mess--nothing elegant (yet).

Consequently, I would like to propose that we track this specific cleanup request as an enhancement.

andyross reviewed (commented):
Finally had a chance to give this a more serious look.  Some notes and nitpicks, none fatal.  The only one that we need to address before merge is the collision with CONFIG_KERNEL_COHERENCE, as that's going to cause test failures on intel_adsp (or rather, if it doesn't then we have a problem with pipe test coverage on SMP, heh).

And the design note about merging the two "zipper" loops for the copies seems really attractive to me.  This PR is doing so much otherwise to symmetrize the zero-copy implementation that it would really be a shame not to finish the job (though it's also possible I'm just mistaken and that won't work).

peter-mitsis commented on kernel/pipes.c:
```diff
@@ -392,147 +290,168 @@ static int pipe_return_code(size_t min_xfer, size_t bytes_remaining,
-		desc->bytes_to_xfer -= bytes_copied;
-	}
+	*bytes_written = pipe_write(pipe, &src_list,
+				    &dest_list, &reschedule_needed);
```
Yes, it is doing it all inside a spinlock. There definitely are latency costs when transferring large amounts of data at once. I thought that I had made mention of this somewhere, but perhaps that was an earlier draft. I should update the documentation to make note of this.

The source of this stems from the initial design decision that is systemic throughout the Zephyr kernel--if someone is waiting on a resource when it is made available, we assign it immediately to the first waiter(s).

peter-mitsis commented on kernel/pipes.c:
```diff
@@ -557,173 +476,156 @@ static int pipe_get_internal(k_spinlock_key_t key, struct k_pipe *pipe,
-	 */
+	src_desc = (struct k_pipe_desc *)sys_dlist_get(&src_list);
+	while (src_desc != NULL) {
+		bytes_copied = pipe_xfer(dest_desc.buffer,
```
I can indeed do this.

peter-mitsis commented on kernel/pipes.c:
```diff
@@ -22,8 +22,10 @@
 
 struct k_pipe_desc {
+	sys_dnode_t    node;
 	unsigned char *buffer;           /* Position in src/dest buffer */
```
Each blocked reader has a buffer into which data will be copied. Each blocked writer has a buffer from which data will be copied. What each blocked reader/writer needs to know is where to/from copy the data and how much of it is left to copy. As the entire setup+copy is within a spinlock, there should not be a higher-priority thread sneaking in to do either a read or a write. This does come at the price--a longer interrupt latency on large transfers.

peter-mitsis commented on kernel/pipes.c:
```diff
@@ -392,147 +290,168 @@ static int pipe_return_code(size_t min_xfer, size_t bytes_remaining,
+						    &pipe->wait_q.readers,
+						    bytes_to_write);
+
+	if (pipe->bytes_used != pipe->size) {
```
I'll revisit this.

peter-mitsis replied:
I've been thinking about this some more and I am sure that this is still the right thing.

The routine pipe_buffer_list_populate() is used by both z_impl_k_pipe_put() and pipe_get_internal() to create a linked list of copy-work to do. Those two functions have one key difference in how the routine is used. When writing to the pipe, we do not want to call it when the pipe is full. When reading from the pipe, we do not want to call it when the pipe is empty. The problem is that from an index perspective, a full pipe looks like an empty pipe. That is, both the read and write indices are the same. The only way to currently distinguish between the two is by comparing bytes_used to size.

Without the check (whether internal to pipe_buffer_list_populate() or external to it), we run the risk of either reading a buffer size worth of data that is not there, or over-writing a buffer size worth of data.

peter-mitsis commented on kernel/pipes.c:
```diff
@@ -392,147 +290,168 @@ static int pipe_return_code(size_t min_xfer, size_t bytes_remaining,
-	size_t         num_bytes_written = 0;
-	size_t         bytes_copied;
+	struct k_pipe_desc  pipe_desc[2];
+	struct k_pipe_desc  src_desc;
```
I do have an idea what we can do for this case. We could pre-allocate a set of configurable pipe descriptors. When one is needed, we could then grab it from the list. When we are done with it, we return it back to the list. In the event that the list is empty, an appropriate error would need to be returned. Admittedly, it's not a perfect solution, but it is not without precedent.

peter-mitsis replied:
Or alternatively, we can add a pipe descriptor into the thread structure for KERNEL_COHERENCE=y systems.

peter-mitsis commented on kernel/pipes.c:
```diff
@@ -392,147 +290,168 @@ static int pipe_return_code(size_t min_xfer, size_t bytes_remaining,
-	} else {
-		k_sched_unlock();
-	}
+	z_pend_curr(&pipe->lock, key, &pipe->wait_q.writers, timeout);
```
Can do.

peter-mitsis commented:
Unless I missed something, the failing checks do appear related to the pipes changes.

andyross commented on kernel/Kconfig:
```diff
@@ -910,6 +910,15 @@ config KERNEL_COHERENCE
+	  When enabled, this option allows pipes to be safely used on a
+	  cache-incoherent architecture. If pipes are not being used on such
+	  a system, it is recommended that this option be disabled so that
+	  some small amount of space in 'struct k_thread' can be saved.
```
Not loving this as an API choice.  Clearly the intent is that you want an app author to be able to say "I don't care about pipes" and save a word.  But having made that choice, nothing tells future authors about it.  The effect is that you end up with a configuration where the pipe API works, mostly, except across CPUs in subtle and weirdly broken ways. 

I think if you want to make this optional, it's probably better to just have "config KERNEL_PIPE" to en/disable the whole API rather than have this thing that ties two otherwise unrelated kconfigs together.

peter-mitsis replied:
It is easy enough to alter things to make pipes a configurable element and have them always use the 'pipe descriptor' stored in the thread. I have no objections to doing that.

andyross reviewed (commented):
A whine about the kconfig choice that rises just above my NAK threshold.  :)

Also, it would definitely be good to see the docs for k_pipe_put/get() updated to explicitly call out the latency effects.  The block copies are done with a spinlock held, so callers need to know not to try to stuff kb-scale data down in one call in latency-critical apps.

peter-mitsis committed:
tests: Add CONFIG_PIPES to tests that use pipes

Use of pipes is now configurable. All tests that use pipes must enable
that feature. (Note: no sample projects currently use pipes.)

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

andyross commented on kernel/Kconfig:
```diff
@@ -554,6 +554,13 @@ config EVENTS
 	  thread structure.
 
+config PIPES
+	bool "Pipe objects"
```
Can we get a "depends on !KERNEL_COHERENCE" here, or a `BUILD_ASSERT(!defined(CONFIG_KERNEL_COHERENCE));` in code somewhere, or some other equivalent validation (with an explanation somewhere adjacent) just so no one trips over the issue accidentally?

andyross reviewed (commented):
One last nitpick, I promise :)

andyross reviewed (dismissed):
Oh wait, I see it.  The pipe record is now always in the thread struct, so it's always compatible.  Never mind, looks great.

andyross commented:
Note that this kind of "record used by a thread only when it's pended" thing is something other areas might productively need to (there's something similar for k_poll vs. COHERENCE IIRC).  Might be useful to have a framework for unioning them together to save space.

peter-mitsis commented:
This latest set of changes fixes a build error found during the last set of checks.  When CONFIG_NET_SOCKETPAIR is defined, it drags in pipe code, which in turn requires CONFIG_PIPES to be defined.

andyross reviewed (approved):
Refresh +1
